### PR TITLE
[WIP] Add provision support as a simplenfs plugin

### LINKED ISSUE #11: Support CreateVolume or DeleteVolume
previously, I use nfs-client-provisioner from [repo](https://github.com/kubernetes-incubator/external-storage/blob/master/nfs-client/cmd/nfs-client-provisioner/provisioner.go) .
Except from the size, It works fine because it can provision and delete. 
So I think if csi-driver-nfs should add this feature?
If yes, I can create a PR  or discuss more detail

### PR DESCRIPTION
**Pull request by mkimuram**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds provision support to csi-driver-nfs.
By adding this feature, user will be able to provision PV with CSI nfs driver, as it could be done with NFS-Client Provisioner for in-tree driver.

**Which issue(s) this PR fixes**:
Fixes #11

**Special notes for your reviewer**:
This PR adds provision feature by using golang's plugin feature. (Also see [here](https://golang.org/pkg/plugin/) and [here](https://medium.com/learning-the-go-programming-language/writing-modular-go-programs-with-plugins-ec46381ee1a9))
The reason that it is implemented by this way is to make a room for nfs vendors to implement their own provision logic as a plugin. (Just building their plugin.so and load it by specifying "--controllerPlugin=" parameter to controller's nfs-plugin will make their code run as a part of csi-driver-nfs).
This will be the similar separation of provision related logics and attach related logics, as it was done in external provisioner and in-tree driver. So, the same concept is expected to be leveraged by iSCSI and FC csi-drivers.

One drawback of using golang's plugin feature is that it doesn't seem to allow build with CGO_ENABLED=0, as a result, it might not support other architectures like arm. Therefore, we might consider if this approach is good or not by taking it into account. (Hopefully, it can be resolved somehow, technically.)

**Does this PR introduce a user-facing change?**:
```release-note

```

### CODE REVIEW

**Comment by mkimuram** [comment]
I would like to focus a discussion on whether we should use golang's plugin feature to implement provision feature on CSI plugins, first. Any feedback is welcomed.

@vladimirvivien 
I was so much inspired by your blog entry to implement this. Thanks for it!
So, I guess you will have feedback on this.


Note that this feature could be tested by below way. (Only storageclass.yaml will require modification.)
<details>

```
# kubectl create -f deploy/kubernetes-simplenfs/csi-attacher-rbac.yaml 
# kubectl create -f deploy/kubernetes-simplenfs/csi-nodeplugin-rbac.yaml 
# kubectl create -f deploy/kubernetes-simplenfs/csi-attacher-nfsplugin.yaml
# kubectl create -f deploy/kubernetes-simplenfs/csi-nodeplugin-nfsplugin.yaml

# kubectl get pod
NAME                             READY   STATUS    RESTARTS   AGE
csi-attacher-nfsplugin-0         3/3     Running   0          45s
csi-nodeplugin-nfsplugin-dtbth   2/2     Running   0          28s

# kubectl create -f examples/kubernetes-simplenfs/storageclass.yaml 
# kubectl create -f examples/kubernetes-simplenfs/pvc.yaml
# kubectl create -f examples/kubernetes-simplenfs/nginx.yaml

# kubectl get pod,pv,pvc
NAME                                 READY   STATUS    RESTARTS   AGE
pod/csi-attacher-nfsplugin-0         3/3     Running   0          4m33s
pod/csi-nodeplugin-nfsplugin-dtbth   2/2     Running   0          4m16s
pod/nginx-simplenfsplugin            1/1     Running   0          29s

NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                          STORAGECLASS       REASON   AGE
persistentvolume/pvc-dd3f788f-72ec-4445-995f-ac98f781304c   1Gi        RWX            Delete           Bound    default/data-simplenfsplugin   csi-simplenfs-sc            2m27s

NAME                                         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
persistentvolumeclaim/data-simplenfsplugin   Bound    pvc-dd3f788f-72ec-4445-995f-ac98f781304c   1Gi        RWX            csi-simplenfs-sc   2m27s
```

File: pkg/nfs/controllerserver.go [hunk 1]
```diff
@@ -12,11 +14,51 @@ type ControllerServer struct {
...
+		}
+		return symbol, nil
+	}
+	return nil, nil
+}
+
 func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	glog.Infof("CreateVolume called")
+	symbol, err := lookupSymbol(cs.Driver.controllerPlugin, "CreateVolume")
+	if err == nil && symbol != nil {
```
[the remark this answers is not in the document]
Comment (mkimuram) [reply pkg/nfs/controllerserver.go]: Thanks! I will fix it.

File: pkg/nfs/nfs.go [hunk 2]
```diff
@@ -42,21 +43,29 @@ var (
...
 	}
 
 	n.AddVolumeCapabilityAccessModes([]csi.VolumeCapability_AccessMode_Mode{csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER})
-	// NFS plugin does not support ControllerServiceCapability now.
-	// If support is added, it should set to appropriate
-	// ControllerServiceCapability RPC types.
-	n.AddControllerServiceCapabilities([]csi.ControllerServiceCapability_RPC_Type{csi.ControllerServiceCapability_RPC_UNKNOWN})
+	glog.Infof("controllerPlugin: %s", n.controllerPlugin)
+	glog.Infof("CreateVolume: %v, DeleteVolume: %v", isSupported(n.controllerPlugin, "CreateVolume"), isSupported(n.controllerPlugin, "DeleteVolume"))
+	createVolume, _ := lookupSymbol(n.controllerPlugin, "CreateVolume")
```
[the remark this answers is not in the document]
Comment (mkimuram) [reply pkg/nfs/nfs.go]: Oh, this is just a debug message, so I will remove these messages later.
(It was needed to find out csi-driver-nfs should be compiled with `CGO_ENABLED=0` to make go plugin work.)

File: simplenfs/plugin.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	"google.golang.org/grpc/status"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/kubernetes/pkg/util/mount"
+)
+
+const (
+	mountPathBase = "/csi-nfs-volume"
+)
+
+func CreateVolume(cs *nfs.ControllerServer, ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
```
[the remark this answers is not in the document]
Comment (mkimuram) [reply simplenfs/plugin.go]: Thanks. I will fix it.

**Comment by mkimuram** [comment]
@vladimirvivien 

I've improved plugin to be loaded as interface. Could you review the code again if this is what you meant.

In this version, features like snapshot and volume expand are not implemented in plugin side, but are implemented in controller side, so once these features are implemented in plugin side, they should work as expected.

File: pkg/nfs/nfs.go [hunk 4]
```diff
@@ -32,17 +35,83 @@ type nfsDriver struct {
...
+		return nil, csc, err
+	}
+
+	// Check if csPlugin implements each capability and add it to implenentation
+	if _, ok := csPlugin.(ControllerPlugin); !ok {
+		glog.Infof("Plugin doesn't implement mandatory methods for controller")
+		return nil, csc, fmt.Errorf("Plugin doesn't implement mandatory methods for controller")
+	}
+
+	if _, ok := csPlugin.(CreateDeleteVolumeControllerPlugin); ok {
```
Comment (vladimirvivien) [inline pkg/nfs/nfs.go]: I would consider changing the following chain to of `if` statements to a type switch:

```go
switch csPlugin.(type) {
   case CreateDeleteVolumeControllerPlugin:
       csc = append(csc, csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME)
   case PublishUnpublishVolumeControllerPlugin:
   ...
   default:
}
```

**Pull request closed without merging**